src/repository/x509.rs

```diff
@@ -103,94 +91,110 @@ impl Name {
-                        while let Some(()) = cons.take_opt_sequence(|cons| {
-                            let id = Oid::take_from(cons)?;
-                            if id == oid::AT_COMMON_NAME {
-                                if cn {
-                                    xerr!(
-                                        return Err(decode::Error::Malformed)
-                                    )
-                                }
-                                Self::skip_router_string(cons)?;
-                                cn = true;
+    ) -> Result<(), InspectionError> {
+        fn inspect_strict<S: decode::Source>(
+            cons: &mut decode::Constructed<S>
+        ) -> Result<(), DecodeError<S::Error>> {
+            let mut cn = false;
+            let mut sn = false;
+            cons.take_sequence(|cons| {
+                while let Some(()) = cons.take_opt_set(|cons| {
+                    while let Some(()) = cons.take_opt_sequence(|cons| {
+                        let id = Oid::take_from(cons)?;
+                        if id == oid::AT_COMMON_NAME {
+                            if cn {
+                                return Err(cons.content_err(
+                                    "invalid router subject"
+                                ))
                             }
-                            else if id == oid::AT_SERIAL_NUMBER {
-                                if sn {
-                                    xerr!(
-                                        return Err(decode::Error::Malformed)
-                                    )
-                                }
-                                Self::skip_router_string(cons)?;
-                                sn = true;
+                            Name::skip_router_string(cons)?;
+                            cn = true;
+                        }
+                        else if id == oid::AT_SERIAL_NUMBER {
+                            if sn {
+                                return Err(cons.content_err(
+                                    "invalid router subject"
+                                ))
                             }
-                            Ok(())
-                        })? { }
+                            Name::skip_router_string(cons)?;
+                            sn = true;
+                        }
                         Ok(())
-                    })? {}
-                    Ok(())
-                })?;
-                // Common name is MUST.
-                if cn {
+                    })? { }
                     Ok(())
-                }
-                else {
-                    Err(decode::Error::Malformed)
-                }
+                })? {}
+                Ok(())
+            })?;
+            // Common name is MUST.
+            if cn {
+                Ok(())
+            }
+            else {
+                Err(cons.content_err("invalid router subject"))
```

Comment:
Missing common name? (also this is not the SN, this is testing the CN, so is subject the right term here?)